raster/r.univar/r.univar_main.c

```diff
@@ -379,7 +409,215 @@ process_raster(univar_stat * stats, int fd, int fdz, const struct Cell_head *reg
 	if (!(param.shell_style->answer))
 	    G_percent(row, rows, 2);
     }
+    G_free(raster_row);
+    if (n_zones)
+        G_free(zoneraster_row);
     if (!(param.shell_style->answer))
 	G_percent(rows, rows, 2);	/* finish it off */
 
 }
+
+#if defined(_OPENMP)
+static void
+process_raster_threaded(univar_stat * stats, char *fname, char *zone_fname, const struct Cell_head *region, const int threads)
+{
+    /* use G_window_rows(), G_window_cols() here? */
+    const unsigned int rows = region->rows;
+    const unsigned int cols = region->cols;
+    const int n_zones = zone_info.n_zones;
+
+    const char *mapset;
+    const char *zone_mapset;
+    mapset = G_find_raster2(fname, "");
+    if (n_zones) {
+        zone_mapset = G_find_raster2(zone_fname, "");
+    }
+
+    /* open file per thread */
+    int *fd, *fdz;
+    fd = G_malloc(threads * sizeof(int));
+    if (n_zones) {
+        fdz = G_malloc(threads * sizeof(int));
+    }
+    for (int i = 0; i < threads; i++) {
+        fd[i] = Rast_open_old(fname, mapset);
+        if (n_zones)
+            fdz[i] = Rast_open_old(zone_fname, zone_mapset);
+    }
+
+    const RASTER_MAP_TYPE map_type = Rast_get_map_type(fd[0]);
+    const size_t value_sz = Rast_cell_size(map_type);
+    void **raster_row;
+    CELL **zoneraster_row = NULL;
+    raster_row = G_malloc(threads * sizeof(void*));
+    if (n_zones) {
+        zoneraster_row = G_malloc(threads * sizeof(CELL*));
+    }
+    for (int i = 0; i < threads; i++) {
+        if (NULL == (raster_row[i] = Rast_allocate_buf(map_type))) {
+            return;
+        }
+        if (n_zones != 0 && (NULL == (zoneraster_row[i] = Rast_allocate_c_buf()))) {
+            return;
+        }
+    }
+
+    int n_alloc = n_zones ? n_zones : 1;
+    size_t **n = G_malloc(threads * sizeof(size_t*));
+    double **sum = G_malloc(threads * sizeof(double*));
+    double **sumsq = G_malloc(threads * sizeof(double*));
+    double **sum_abs = G_malloc(threads * sizeof(double*));
+    double **min = G_malloc(threads * sizeof(double*));
+    double **max = G_malloc(threads * sizeof(double*));
+    size_t **size = G_malloc(threads * sizeof(double*));
+
+    for (int i = 0; i < threads; i++) {
+        n[i] = G_calloc(n_alloc, sizeof(size_t));
+        sum[i] = G_calloc(n_alloc, sizeof(double));
+        sumsq[i] = G_calloc(n_alloc, sizeof(double));
+        sum_abs[i] = G_calloc(n_alloc, sizeof(double));
+        size[i] = G_calloc((n_alloc), sizeof(size_t));
```

Comment:
There is no need for round brackets around n_alloc.

Reply:
Alright, will fix that.